Vector.c

```diff
@@ -61,6 +74,9 @@ static bool Vector_isConsistent(const Vector* this) {
 }
 
 unsigned int Vector_count(const Vector* this) {
+   if (Vector_isDirty(this)) {
+      return this->items - this->dirty_count;
+   }
```

Comment:
While I know where this comes from, I'm not really happy about this, as this value is potentially used for the upper bound when iterating over the vector, thus hiding the last few items if the vector is iterated from zero to count …

Reply:
Apologies for the slow response and yeah, this isn't great as it could easily lead to confusion. Since this function is only used in assertions by [`ProcessList_add()`](https://github.com/htop-dev/htop/blob/6133cac721f355aec0d6b5581fef90db34305f29/ProcessList.c#L167) and [`ProcessList_remove()`](https://github.com/htop-dev/htop/blob/6133cac721f355aec0d6b5581fef90db34305f29/ProcessList.c#L191) we could rename it to something like `Vector_countNonNull()`, add a comment about its usage, and fail with `CRT_fatalError()` if `NDEBUG` is set.

A more aggressive option would be to define macros for vector iteration and enforce their use.

```c
#define CONCATX(x, y)      x##y
#define CONCAT(x, y)       CONCATX(x, y)
#define UNIQUE_ALIAS(name) CONCAT(name, __COUNTER__)

/*
 * for_each_if - helper for handling conditionals in various for_each macros
 * @condition: The condition to check
 *
 * Typical use::
 *
 * #define for_each_foo_bar(x, y) \'
 *    list_for_each_entry(x, y->list, head) \'
 *       for_each_if(x->something == SOMETHING)
 *
 * The for_each_if() macro makes the use of for_each_foo_bar() less error
 * prone.
 */
#define for_each_if(condition) if (!(condition)) {} else

/*
 * _Vector_forEach - internal macro to iterate over all entries in Vector.
 *
 * @vec: vector to iterate
 * @value: *Object iteration cursor
 * @__items: int iteration limit, for macro-internal use
 * @__i: int iteration cursor, for macro-internal use
 */
#define _Vector_forEach(vec, value, __items, __i)            \
   const int (__items) = Vector_size(vec);                   \
   for (int (__i) = 0; (__i) < (__items); (__i)++)           \
      for_each_if(((value) = Vector_get((vec), (__i))))

/*
 * Vector_forEach - iterate over all entries in Vector
 * @vec: vector to iterate
 * @value: *Object iteration cursor
 */
#define Vector_forEach(vec, value)                           \
   _Vector_forEach(vec, value, UNIQUE_ALIAS(__items), UNIQUE_ALIAS(__i))

/*
 * _Vector_forEachIndex - internal macro to iterate over all entries in Vector
 * @vec: vector to iterate
 * @__items: int iteration limit, for macro-internal use
 * @__i: int iteration cursor, for macro-internal use
 */
#define _Vector_forEachIndex(vec, __items, __i)              \
   const int (__items) = Vector_size(vec);                   \
   for (int (__i) = 0; (__i) < (__items); (__i)++)           \
      for_each_if(((idx) == __i) || true);

/*
 * Vector_forEachIndex - iterate over all entries in Vector
 * @vec: vector to iterate
 * @idx: int iteration cursor
 */
#define Vector_forEachIndex(vec, idx)                                  \
   for (int __i = 0, __items = Vector_size(vec); __i < __items; __i++) \
      for_each_if(((idx) == __i) || true);

```

Reply:
What arguments do we have against simply ignoring the `NULL` elements and just return the number of (`NULL` and non-`NULL`) items?

Having all that macro boilerplate is IMO not helping with making the code clean as it hides quite a bit of control flow.

Reply:
On second thought:

Split `Vector_count` into two routines: `Vector_count` and `Vector_countUnsafe`.

```c
unsigned int Vector_countUnsafe(const Vector* this) {
   return this->items;
}

unsigned int Vector_count(const Vector* this) {
   assert(Vector_isConsistent(this));
   return Vector_countUnsafe(this);
}
```

Rationale: Sticks with the current API of allowing access to any element, but doesn't throw you of if elements where deleted. The `*Unsafe` is for places that expect missing items, thus pointing out, that the following `Vector_get` calls may return `NULL`. On the other hand, code not expecting missing items will work as is and barf about the inconsistency in development builds. You could even drive this further by Introducing a `Vector_getUnsafe` function with the same semantics. Thus any code not expecting `NULL` values could throw a fatal error upon encountering such invalid indices.

Reply:
Taking another look at this, `Vector_count()` is only used in two places: [`ProcessList_add()`](https://github.com/htop-dev/htop/blob/6133cac721f355aec0d6b5581fef90db34305f29/ProcessList.c#L167) and [`ProcessList_remove()`](https://github.com/htop-dev/htop/blob/6133cac721f355aec0d6b5581fef90db34305f29/ProcessList.c#L191) (to assert that the hash table and vector contain the same number of elements). And the concern is that future devs might use this as an upper bound when iterating over a vector, why don't we remove this function and replace it with something like `Vector_assertCountEquals()` that asserts the vector contains `N` non-null items? This would remove the possibility of confusing this with `Vector_size()`.

Reply:
Also, is this actually an issue since [`Vector_count()`](https://github.com/htop-dev/htop/blob/c4239335b9fb2da6dab1f53ff0e8eb32c78530de/Vector.h#L53-L57) is only defined when `NDEBUG` is set?

Regardless, I replaced `Vector_count` with `Vector_countEquals` in this commit: https://github.com/htop-dev/htop/pull/939/commits/79971f4980beee1086b0316f243e8ea2a191e53c. If you like this commit I'll squash it, otherwise we can revert it and find another solution.

Reply:
The commit message of that commit misses one negation. `Vector_count` is only define if `NDEBUG` is not defined. Also the sentence in that commit message might profit from some grammar simplification. Apart from this I think that renaming is an improvement. Will have to concentrated look at the overall PR again, to finalize my view on this. But I think we're getting close …

Reply:
Sounds good and I'll fixup that commit message when I rebase/squash it.

Reply:
Squashed the commit message